Add background job to clean up orphaned properties

sharidas opened the pull request:
Add a background job to clean up orphaned properties
in the table.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" in case the PR still has open tasks
- Set label "backport-request" if backport is needed
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
<!--- Describe your changes in detail -->
Clean up orphan properties in the table.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes https://github.com/owncloud/core/issues/33413

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Clean up orphan properties.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Create files a.txt, b.txt, c.txt and d.txt
- Add properties to all the files created above, by using proppatch.xml file `<?xml version="1.0" encoding="utf-8" ?><propertyupdate xmlns='DAV:'><set><prop><t:valnspace xmlns:t='http://example.com/neon/litmus/'><foo xmlns='http://bar'/></t:valnspace></prop></set></propertyupdate>`
- Verify that oc_properties table has been updated
- Now delete the file `a.txt` and `c.txt` using a small script 
```<?php
require_once __DIR__ . '/lib/base.php';
/*$a = new \OC\Files\View('/admin/files/');
$a->unlink('a.txt');*/
//$name = \OC::$server->getRootFolder()->getName('/admin/files/a.txt');
$adminUser = \OC::$server->getUserFolder('admin');
$a = $adminUser->get('a.txt');
$a->delete();
```
- Verify there are 4 entries in the oc_properties.
- Made small modification to the code by adding method in CleanProperties.php
```
	public function execute($jobList, ILogger $logger = null) {
		$this->processProperties();
	}

```
- Now run the background job as ```./occ system:cron```.
- Verify that there should be only 2 entries. And the fileid's of oc_properties should belong to fileid's of `b.txt` and `d.txt.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

sharidas committed:
Add background job to clean up orphaned properties

Add a background job to clean up orphaned properties
in the table.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
The essence of this PR,
- It tries to iterate through the properties table and fetches the fileid
- It checks if this fileid is there in the filecache table.
- If fileid is found in filecache, it wont delete the property from the table, else the property is deleted.
Let me know if this change looks ok. @PVince81

micbar commented on apps/dav/lib/BackgroundJob/CleanProperties.php:
```diff
@@ -0,0 +1,116 @@
+	 */
+	private function getAllFileIds() {
+		$qb = $this->connection->getQueryBuilder();
+		$qb->select('fileid')
```
Would be interesting for Files Lifcycle too. Fo the moment, we have a SQL Query like this 
https://github.com/owncloud/files_lifecycle/blob/master/lib/ArchiveQuery.php#L44

I have a WIP query for the query builder, which doesn't work.
https://gist.github.com/micbar/95e7a7edd9eda5c391b300da9b96a0f2

It gives me a SQL syntax error. I suspect, that the escaping of the property names is the problem here https://gist.github.com/micbar/95e7a7edd9eda5c391b300da9b96a0f2#file-archivequery-php-L18

sharidas commented:
Current state of the PR:
- [x] Updated the query to use a left join. The query will provide the fileid's to be deleted from the oc_properties table. And then delete it.

Kindly review this PR.

PVince81 commented on apps/dav/lib/BackgroundJob/CleanProperties.php:
```diff
@@ -0,0 +1,94 @@
+		$result = $qb->execute();
+
+		while ($row = $result->fetch()) {
+			$this->deleteOrphan($row['fileid']);
```
deleting one by one could be slow. might be better to batch them in chunks and have the deletion query use `where fileid in (...)`

PVince81 replied:
did you try `delete from oc_properties p where p.fileid not in (select fileid from oc_filecache f)` ?
as far as I remember this probably doesn't work on all databases so your current approach is closer to the better compromise

sharidas replied:
> did you try `delete from oc_properties p where p.fileid not in (select fileid from oc_filecache f)` ?
> as far as I remember this probably doesn't work on all databases so your current approach is closer to the better compromise

No I have not done this way.

PVince81 reviewed (commented):
Looks better.
A little perf improvement would be nice, see comment.

sharidas commented:
Current state of the PR:
- [ ] delete operation is not done one by one. Its done as per the suggestion `where fileid in (...)`. I also preferred to delete in a batch of 200 fileids.
- [ ] Updated the tests for large deletion ( 300 entries deleted from 310 entries ).

sharidas commented:
Oops postgresql failed. 
- [ ] Verify the change works with postgresql.

sharidas commented:
@PVince81 another round of review is required. 

- I have updated the unit test for the postgresql failure.
- Also points mentioned at https://github.com/owncloud/core/pull/33464#issuecomment-441184672 remains valid for the review.

PVince81 reviewed (commented):
See comment. The chunking solution should also cover the detection query.

sharidas commented on apps/dav/lib/BackgroundJob/CleanProperties.php:
```diff
@@ -0,0 +1,114 @@
+				return (string) $val;
+			}, $items);
+
+			$this->logger->info("Deleting property for fileids: " . \implode(',', $strFileIds), ['app' => 'dav']);
```
Updated the code to use `debug` and the changeset now shows the total number of entries deleted.

> do we use `info` in other background jobs ? as far as I remember we only summarize by saying how many entries we processed, not more

Nope. I thought since its an information to the admin, hence added info. Looked into some of the background jobs they do have debug. And they just summarize saying how many entries were deleted.

sharidas commented:
The current state of the PR:

- [x] Updated the change set to use `debug`, instead of `info` to log. Also the log will have only the total number of deleted entries.
- [x] Updated the select query to provide 200 entries each time. And these entries will be passed on to the delete query, until nothing is left.

sharidas commented on apps/dav/lib/BackgroundJob/CleanProperties.php:
```diff
@@ -0,0 +1,114 @@
+
+		$fileIds = [];
+		while ($row = $result->fetch()) {
+			$fileIds[] = (int) $row['fileid'];
```
I have updated the query in the latest version of the PR.

PVince81 reviewed (approved):
:+1:

PVince81 commented:
@sharidas please backport

sharidas commented:
Backport to stable10 https://github.com/owncloud/core/pull/33677
